src/par_schedule.rs

```diff
@@ -0,0 +1,187 @@
+        }
+    }
+}
+
+struct ScheduleWorker<A: 'static +  Agent + Clone >{
+    job_sender: std::sync::mpsc::Sender<Vec<Pair<A>>>,
+    ris_receiver: std::sync::mpsc::Receiver<Vec<Pair<A>>>,
+}
+
+impl <A: 'static +  Agent + Clone + Send + Sync> ScheduleWorker<A>{
+
+
+    fn new()->ScheduleWorker<A>{
+        let (job_s,job_r) = std::sync::mpsc::channel();
+        let (ris_s,ris_r) = std::sync::mpsc::channel();
+        std::thread::spawn(move ||{
+            loop{
+                let jobs : Vec<Pair<A>> = match job_r.recv(){
+                    Ok(i)=> i,
+                    Err(_) => break
+                };
+                let mut ris = Vec::with_capacity(jobs.len());
+                // eprintln!("{:?} ricevuti {} jobs",std::thread::current().id(), jobs.len());
+                for mut item in jobs{
+                    item.agentimpl.step();
+                    if item.agentimpl.repeating{
+                        ris.push(item);
+                    }
+                }
+                ris_s.send(ris).unwrap();
+            }
+        });
+        ScheduleWorker{
+            job_sender: job_s,
+            ris_receiver: ris_r
+        }
+    }
+
+    //sends an array of agents to this ScheduleWorker associated thread 
+    fn execute(&self,jobs: Vec<Pair<A>>){
+        self.job_sender.send(jobs).unwrap();
+    }
+
+    //returns repeating agents from previous execute() call
+    fn take(&self) -> Vec<Pair<A>>{
+        self.ris_receiver.recv().unwrap()
+    }
+
+}
+
+impl<A: 'static +  Agent + Clone + Send + Sync > Schedule<A> {
+
+    pub fn new() -> Schedule<A> {
+        Schedule {
+            step: 0,
+            time: 0.0,
+            events: Mutex::new(PriorityQueue::new()),
+            workers : Vec::with_capacity(THREAD_NUM)
+        }
+    }
+
+    pub fn schedule_once(&self, agent: AgentImpl<A>,the_time:f64, the_ordering:i64) {
+        self.events.lock().unwrap().push(agent, Priority{time: the_time, ordering: the_ordering});
+    }
+
+    pub fn schedule_repeating(&self, agent: A, the_time:f64, the_ordering:i64) {
+        let mut a = AgentImpl::new(agent);
+        a.repeating = true;
+        let pr = Priority::new(the_time, the_ordering);
+        self.events.lock().unwrap().push(a, pr);
+    }
+
+    pub fn simulate<S: State>(&mut self, state: &S, num_step:u128){
+        for _ in 0..num_step{
+            self.step(state);
+        }
+    }
+
+    pub fn step(&mut self, state: &impl State){
+       
+        if self.step == 0{
+            state.update();
+            for _ in 0..THREAD_NUM{
+                self.workers.push(ScheduleWorker::new());
+            }
+            
+        }
+        self.step += 1;
+       
+
+        let events = &mut self.events;
+        if events.lock().unwrap().is_empty() {
+            println!("coda eventi vuota");
+            return
+        }
+
+        let thread_division = (events.lock().unwrap().len() as f64 /THREAD_NUM as f64).ceil() as usize ;
```

Comment:
Please explain this division.
I suppose that the best option to divide a work of N is:
```
int count = N / size;
int remainder = N % size;
int start, stop;

if (thread_id < remainder) {
    start = rank * (count + 1);
    stop = start + count;
} else {
    start = rank * count + remainder;
    stop = start + (count - 1);
}
```
using this code, _remainder_ threads take one agent more...

Reply:
Supposing we're using THREAD_NUM threads and N events to process, I'm allocating space for THREAD_NUM vectors each containing at most `thread_division` events.
Later, I'm pushing events into each of these vectors as they're extracted from the PriorityQueue: 
```
let mut i = 0;
loop {
            //checks

            let item = events.lock().unwrap().pop();
            match item {
                Some(item) => {
                    let (agent, priority) = item;
                    cevents[i%THREAD_NUM].push(Pair::new(agent, priority));
                    i+=1;
                },
                None => panic!("no item"),
            }
        }
```
Since I used ceil(), if `remainder>0`, each vector will be able to hold one additional event. With the current solution only the first `remainder` vectors get to make use of this additional space, therefore I should optimize this section.